crates/bifrost/src/types.rs

```diff
@@ -15,29 +15,45 @@ use crate::loglet::LogletOffset;
 use restate_types::logs::{Lsn, SequenceNumber};
 use serde::{Deserialize, Serialize};
 
+// Only implemented for LSNs
 pub(crate) trait LsnExt: SequenceNumber {
     /// Converts a loglet offset into the virtual address (LSN).
+    #[track_caller]
     fn offset_by<S: SequenceNumber>(self, offset: S) -> Self {
         // This assumes that this will not overflow. That's not guaranteed to always be the
         // case but it should be extremely rare that it'd be okay to just wrap in this case.
         //
-        // We subtract from OLDEST because loglets start their offset from 1, not 0.
+        // We subtract from OLDEST because loglets might start offsets from a non-zero value.
         // 1 is the oldest valid offset within a loglet, 0 is an invalid offset.
         debug_assert!(offset != S::INVALID);
         let self_raw: u64 = self.into();
         let offset_raw: u64 = offset.into();
 
-        Self::from(self_raw.wrapping_add(offset_raw) - S::OLDEST.into())
+        Self::from(std::cmp::max(
+            Self::OLDEST.into(),
+            self_raw
+                .wrapping_add(offset_raw)
```

Comment:
Wondering whether we should allow wrapping or instead panic in case of an overflow. Hard to wrap my head around what kind of errors can happen with wrapping. Especially the combination of `r = a.wrapping_add(b).saturating_sub(c)` seems to behave oddly:

a = u64::MAX, b = 0, c = u64::MAX/2 => r = u64::MAX/2 + 1
a = u64::MAX, b = 1, c = u64::MAX/2 => r = 0